src/storage/v2/constraints/type_constraints.hpp

```diff
@@ -0,0 +1,61 @@
+// Copyright 2024 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include <storage/v2/constraints/type_constraints_type.hpp>
+#include <utility>
+#include "storage/v2/constraints/constraint_violation.hpp"
+#include "storage/v2/durability/recovery_type.hpp"
+#include "storage/v2/id_types.hpp"
+#include "storage/v2/vertex.hpp"
+#include "utils/skip_list.hpp"
+
+namespace memgraph::storage {
+
+class TypeConstraints {
+ public:
+  struct MultipleThreadsConstraintValidation {
+    std::optional<ConstraintViolation> operator()(const utils::SkipList<Vertex>::Accessor &vertices,
+                                                  const LabelId &label, const PropertyId &property);
+
+    const durability::ParallelizedSchemaCreationInfo &parallel_exec_info;
+  };
+  struct SingleThreadConstraintValidation {
+    std::optional<ConstraintViolation> operator()(const utils::SkipList<Vertex>::Accessor &vertices,
+                                                  const LabelId &label, const PropertyId &property);
+  };
+
+  [[nodiscard]] static std::optional<ConstraintViolation> ValidateVertexOnConstraint(const Vertex &vertex,
+                                                                                     LabelId label, PropertyId property,
+                                                                                     TypeConstraintsType type);
+
+  [[nodiscard]] static std::optional<ConstraintViolation> ValidateVerticesOnConstraint(
+      utils::SkipList<Vertex>::Accessor vertices, LabelId label, PropertyId property, TypeConstraintsType type);
+
+  [[nodiscard]] std::optional<ConstraintViolation> Validate(const Vertex &vertex);
+  [[nodiscard]] std::optional<ConstraintViolation> Validate(const Vertex &vertex, LabelId label);
+  [[nodiscard]] std::optional<ConstraintViolation> Validate(const Vertex &vertex, PropertyId property);
+
+  bool HasTypeConstraints() const;
+  bool ConstraintExists(LabelId label, PropertyId property) const;
+  bool InsertConstraint(LabelId label, PropertyId property, TypeConstraintsType type);
+  bool DropConstraint(LabelId label, PropertyId property, TypeConstraintsType type);
+
+  std::vector<std::tuple<LabelId, PropertyId, TypeConstraintsType>> ListConstraints() const;
+  void DropGraphClearConstraints();
+
+ private:
+  // TODO Ivan: unordereded set doesn't work, spaceship deleted, i dont know check later
+  std::map<std::pair<LabelId, PropertyId>, TypeConstraintsType> constraints_;
```

Comment:
To facilitate other operation maybe worth duplicating information in multiple forms 
Thinking `map<Label, map<Prop, TypeConstraintsType>>` and `map<Prop, map<Label, TypeConstraintsType>>`

This comes does to a question on performance, need to evaluate cost.